File: internal/tui/chat/thinking.go
package chat

import (
	"fmt"
	"strings"
	"time"

	"github.com/charmbracelet/lipgloss"

	"github.com/georgebuilds/carlos/internal/agent"
)

// Activity indicator ("carlos is thinking") for the gap between a
// user submission and the first model output.
//
// Why this exists: provider RTT plus first-token latency is typically
// 300ms – a few seconds. With nothing on-screen during that window the
// chat reads as frozen, and when several tool calls land in close
// succession they appear "all at once" instead of feeling streamed.
// A subtle live row anchors the wait, signals the model is working,
// and (once elapsed > 3s) reports how long the wait has been.
//
// Visual choice: a 3-dot bouncing-ball pattern is the most legible
// "typing indicator" idiom and avoids the cute-but-distracting word
// games some other CLIs use. Animation runs off the existing 33 ms
// textTick so we don't add another timer.

// thinkingFrames spells out the 3-dot pattern. Each frame is a
// bitmask: true = highlighted dot, false = dim dot. Four frames
// cycle bright dot left → middle → right → middle, which reads as
// a smooth wave.
var thinkingFrames = [][3]bool{
	{true, false, false},
	{false, true, false},
	{false, false, true},
	{false, true, false},
}

// thinkingFrameTicks is the number of textTicks (33 ms each) between
// animation frames. 5 ticks ≈ 165 ms per frame, ~660 ms per full
// cycle — slow enough not to feel jittery, fast enough to feel alive.
const thinkingFrameTicks = 5

// thinkingElapsedThreshold gates the "(Ns)" trailing timer so a quick
// reply doesn't get a stopwatch slapped on it. Once a wait crosses
// this threshold the timer kicks in and updates every render.
const thinkingElapsedThreshold = 3 * time.Second

// isThinking decides whether the chat should paint the activity row
// at the bottom of the transcript. Conditions:
//
//   - The projection reports an "in-flight" state — Spawning (early,
//     before the first state_change lands), Running, or Compacting.
//   - No live assistant text is buffered. Streaming text is its own
//     "alive" signal and the spinner would compete with it.
//   - There IS something in the transcript already (no spinner over a
//     blank welcome screen) AND the last entry is something the
//     model is responding to — a user message, tool call/result, or
//     user-shell command. We don't paint a spinner under a finished
//     assistant turn; that reads as if carlos is stalling.
//   - The chat isn't in research-progress mode — that row is its own
//     activity indicator.
func (m *Model) isThinking() bool {
	state, _ := m.headerState()
	switch state {
	case agent.StateSpawning, agent.StateRunning, agent.StateCompacting:
		// in-flight
	default:
		return false
	}
	if m.source.Get(m.agentID) != "" {
		return false
	}
	if len(m.transcript) == 0 {
		return false
	}
	last := m.transcript[len(m.transcript)-1]
	switch last.kind {
	case entryUserMessage, entryToolCall, entryToolResult, entryUserShell:
		return true
	}
	return false
}

// assistantBusy returns true when the assistant is mid-turn — either
// streaming text into the live buffer, or in one of the in-flight
// projection states (Spawning, Running, Compacting) waiting on a tool
// or model response. Used by submit() to decide whether a fresh
// user-typed line should dispatch immediately or be parked in the
// queue for flushQueuedUserMessage to release once the turn ends.
//
// isThinking() returns false during live text streaming (it defers to
// the streamed text as the alive signal), so OR-ing the two captures
// "the model is currently doing something on our behalf" without
// double-counting either condition.
func (m *Model) assistantBusy() bool {
	if m.source != nil && m.source.Get(m.agentID) != "" {
		return true
	}
	return m.isThinking()
}

// thinkingElapsed returns the wall-clock time since the most recent
// transcript entry. Zero when the transcript is empty. Used by the
// indicator to surface a "(Ns)" trailer once the wait crosses
// [thinkingElapsedThreshold] so the user can gauge whether carlos is
// stalled vs. just slow.
func (m *Model) thinkingElapsed() time.Duration {
	if len(m.transcript) == 0 {
		return 0
	}
	last := m.transcript[len(m.transcript)-1]
	return time.Since(last.ts)
}

// renderThinkingRow paints the activity indicator. Layout:
//
//	🧢   ● ∙ ∙    thinking    (3s)
//
// The bouncing dot is the focal element (accent color, bold); the
// "thinking" label is dim italic so the eye lands on the dots first
// and reads the verb as context. The elapsed timer is hidden under
// 3 s — quick replies stay quiet.
func renderThinkingRow(tick int, elapsed time.Duration, width int) string {
	capStyle := lipgloss.NewStyle().Foreground(colorAgent)
	dimStyle := lipgloss.NewStyle().Foreground(colorSubtle).Italic(true)

	dots := renderThinkingDots(tick)
	label := dimStyle.Render("thinking")

	line := capStyle.Render("🧢") + "   " + dots + "   " + label

	if elapsed >= thinkingElapsedThreshold {
		secs := int(elapsed.Seconds())
		timer := lipgloss.NewStyle().
			Foreground(colorSubtle).
			Render(fmt.Sprintf("   · %ds", secs))
		line += timer
	}

	// Width clamp via lipgloss so a future change to the avatar column
	// doesn't accidentally push the row past the viewport.
	if width > 0 {
		return lipgloss.NewStyle().MaxWidth(width).Render(line)
	}
	return line
}

// renderThinkingDots paints the 3-dot bouncing pattern in one row.
// One dot wears the accent (bright), the other two are subtle so the
// "wave" reads at a glance without forcing the eye to track motion.
func renderThinkingDots(tick int) string {
	frame := (tick / thinkingFrameTicks) % len(thinkingFrames)
	pattern := thinkingFrames[frame]
	bright := lipgloss.NewStyle().Foreground(colorAccent).Bold(true)
	dim := lipgloss.NewStyle().Foreground(colorMuted)
	parts := make([]string, 3)
	for i, on := range pattern {
		if on {
			parts[i] = bright.Render("●")
		} else {
			parts[i] = dim.Render("∙")
		}
	}
	return strings.Join(parts, " ")
}

File: web/src/api/thinking.ts
// carlos web · thinking-indicator derivation.
// Web translation of the TUI activity indicator (internal/tui/chat/thinking.go):
// the gap between a user submission and the first streamed token reads as
// frozen without a sign of life in the reply position. The TUI derives its
// predicate from the projection state plus the transcript tail; here the same
// three questions are asked of the wire-event list and the live delta buffer.
//
//   1. Is the agent in-flight? (latest `state` event, falling back to the
//      roster summary when no state event has landed yet, e.g. the very first
//      send on a fresh thread)
//   2. Is live text already streaming? (the delta buffer is its own alive
//      signal; the dots would compete with it)
//   3. Is the transcript tail something the model is responding to? (a user
//      message, tool call/result, or shell marker. Never an assistant
//      message: dots under a finished turn read as stalling, and an error
//      reply lands as assistant_message too, which is what dismisses the
//      indicator on failure)

import type { WireEvent, WireState } from './types'

// In-flight wire states, mirroring the TUI's Spawning / Running / Compacting
// gate. blocked (approval wait), awaiting_input, failed, done etc. all read
// as "not thinking".
const THINKING_STATES: ReadonlySet<string> = new Set(['spawning', 'running', 'compacting'])

// Transcript kinds that anchor the indicator: the model is responding to
// these. assistant_message is in the set so the tail scan stops on it (and
// then answers "no"); hairline kinds (state, research_phase, children,
// approval_resolved) are skipped so a state event landing after the user
// message does not hide the dots.
const ANCHOR_KINDS: ReadonlySet<string> = new Set([
  'user_message',
  'assistant_message',
  'tool_call',
  'tool_result',
  'shell_start',
  'shell_end',
])

// latestWireState scans backward for the most recent persisted `state` event.
// Null when none has landed (fresh thread before the first transition).
export function latestWireState(events: WireEvent[]): WireState | null {
  for (let i = events.length - 1; i >= 0; i--) {
    if (events[i].kind === 'state') {
      const d = events[i].data as { state?: WireState }
      if (d.state) return d.state
    }
  }
  return null
}

// lastAnchor returns the most recent transcript event the model could be
// responding to (or finishing with), skipping hairline/meta kinds.
export function lastAnchor(events: WireEvent[]): WireEvent | null {
  for (let i = events.length - 1; i >= 0; i--) {
    if (ANCHOR_KINDS.has(events[i].kind)) return events[i]
  }
  return null
}

// isThinking decides whether the transcript should paint the activity row in
// the assistant reply position. Pure over (events, delta, fallbackState) so
// tests drive it directly; callers layer ownership/approval gating on top.
export function isThinking(
  events: WireEvent[],
  delta: string,
  fallbackState?: WireState,
): boolean {
  const state = latestWireState(events) ?? fallbackState ?? null
  if (!state || !THINKING_STATES.has(state)) return false
  if (delta !== '') return false
  const anchor = lastAnchor(events)
  if (!anchor) return false // no dots over a blank transcript
  return anchor.kind !== 'assistant_message'
}
